### ISSUE TITLE
LDAP: cannot login using email

### ISSUE THREAD

**Issue by gregdymek** [go-gitea/gitea#32844]
### Description

I would like to login using my email,
however I'm not able to configure gitea,
User Filter works only when I use (&(objectClass=inetOrgPerson)(cn=%[1]s)), in cn there is my name eg. Greg,
when I first log in using "Greg" then it works the user got synced with his email address and now I'm able to login using my email,
but if I change User Filter to (&(objectClass=inetOrgPerson)(mail=%[1]s)) or (&(objectClass=inetOrgPerson)(|(cn=%[1]s))(mail=%[1]s))) and try to login to gitea without prior User sync it doesn't work,
I'm not sure if there is something I'm doing wrong or there is a bug,
pls help :), 
thanks in advance

### Gitea Version

1.22.6

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

2024/12/14 18:46:22 ...vices/auth/signin.go:119:UserSignIn() [W] Failed to login 'greg@dymek.com' via 'ldap': name is invalid [greg@dymek.com]: must be valid alpha or numeric or dash(-_) or dot characters
2024/12/14 18:46:22 ...ers/web/auth/auth.go:231:SignInPost() [I] Failed authentication attempt for greg@dymek.com from 91.236.87.143:0: Email address does not exist [email: greg@dymek.com]

### How are you running Gitea?

I'm running using docker compose

### Database

PostgreSQL

**Comment by wxiaoguang** [comment]
What's your LDAP result? I guess you used "cn=user@domain.com" or "uid=user@domain.com", then the email address can't be used as the username.

In short: the real user name used by Gitea must be a pure username, but not an email address.

**Comment by gregdymek** [comment]
hi wxiaoguang, thank you for your help,
in cn I have my name - "Greg" in mail I have "greg@dymek.com"
when I try to login with my email the first time I set up gitea it's not possible (or I'm doing something wrong) I have to use whatever I have in "User Filter" in my case it's "Greg",
once I initially logged in as "Greg" and logged out now I'm able to use my email (as the user got synced with gitea),
what I would like to do is to be able to log in using my email without having to login using User Filter first,
I see there is "TestLDAPEmailSignin" test but I think it doesn't cover the scenario (because I assume User Filter is responsible for both authentication and user mapping)
do you think it's possible?
thank you

**Comment by wxiaoguang** [comment]
> I see there is "TestLDAPEmailSignin" test but I think it doesn't cover the scenario (because I assume User Filter is responsible for both authentication and user mapping)

It depends on how your LDAP is configured (`directBind`).

I am also using LDAP in my instance. If I understand correctly: LDAP first tries to "bind" (either by the current userDN&password, or the bindDN&password), then it collects the entity information and Gitea uses the entity information to do login and/or user-creation.

----

Since you have seen `must be valid alpha or numeric or dash(-_) or dot characters` in your error log, it seems that your setup is trying to create a new user with the email address. My tests just show that "login with email could create the user by its name correctly".

So the question is: what's the search result of your instance? Is the "name" the pure username? If you'd like to debug into the problem, you could try to add some logs in https://github.com/go-gitea/gitea/blob/92648112175753c4eaa13cc6f8e3cc6a3cef9ad5/services/auth/source/ldap/source_search.go#L252

**Comment by wxiaoguang** [comment]
Or if you could provide a reproducible setup with detailed steps (for example, a docker compose with necessary gitea&openldap config), and then I can reproduce your problem on my side, then maybe I could also try to figure it out or fix it.

**Comment by wxiaoguang** [comment]
And one more question: what's your "Username Attribute"? If it doesn't exist in the LDAP search result, then Gitea will use the "inputed username (in your case: email)" as the username's fallback. Then you would still see the error above.

![image](https://github.com/user-attachments/assets/74854b2b-ee4a-4c32-a303-1176331fb58c)

**Comment by gregdymek** [comment]
thank you wxiaoguang, I think username attribute is sn, uid is a uuid, but let me try again

**Comment by gregdymek** [comment]
ok, I think I was missing user attribute and what you said happened (it was trying to use email as the username attribute), now it's working, I really appreciate your quick help, thank you again
